From: Teerth Reddy <teerth@ti.com>

[PATCH] Clock Domain test

This patch has the Clock Domain Test.Needs the cd_test.ko module
to invoke this.

Signed-off-by: Teerth Reddy <teerth@ti.com>
---
 arch/arm/mach-omap2/Makefile       |    3 +
 arch/arm/mach-omap2/clk_dm_test.c  |   83 +++++++++++++++++++++++++++++++++++++
 arch/arm/mach-omap2/clockdomains.h |   18 ++++++--
 3 files changed, 100 insertions(+), 4 deletions(-)

Index: kernel-omap3/arch/arm/mach-omap2/clk_dm_test.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ kernel-omap3/arch/arm/mach-omap2/clk_dm_test.c	2009-12-17 13:07:15.000000000 +0530
@@ -0,0 +1,83 @@
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/moduleparam.h>
+#include <linux/file.h>
+#include <asm/uaccess.h>
+#include <linux/mm.h>
+
+#include <plat/powerdomain.h>
+#include <plat/clock.h>
+#include <plat/clockdomain.h>
+
+
+
+
+
+
+extern struct clockdomain *clockdomains_omap[];
+void clock_domain_test(void);
+
+static struct clk dummy_clk1 = {
+	.name = "clk1",
+};
+
+void clock_domain_test()
+{
+	int val, i;
+	static struct clockdomain *cd;
+	static struct powerdomain *pd;
+
+	for (i = 0; clockdomains_omap[i] != NULL; i++) {
+			val = omap2_clkdm_clk_enable(cd = clockdomains_omap[i],
+				 &dummy_clk1);
+		if (val == 0)
+			printk(KERN_INFO "Enabled Clock for %s\n", cd->name);
+		else
+			printk(KERN_ERR "Failed with value %d for %s\n",
+								 val, cd->name);
+	}
+
+	for (i = 0; clockdomains_omap[i] != NULL; i++) {
+		val = omap2_clkdm_clk_disable(cd = clockdomains_omap[i],
+				 &dummy_clk1);
+		if (val == 0)
+			printk(KERN_INFO "Disabled Clock for %s\n", cd->name);
+		else
+			printk(KERN_ERR "Failed with value %d for %s\n",
+								 val, cd->name);
+	}
+
+	for (i = 0; clockdomains_omap[i] != NULL; i++) {
+		pd = clkdm_get_pwrdm(cd = clockdomains_omap[i]);
+		if (pd)
+			printk(KERN_INFO "%s is PWRDM for %s\n", pd->name,
+							 cd->name);
+		else
+			printk(KERN_ERR "Invalid Clock Domain\n");
+	}
+
+	for (i = 0; clockdomains_omap[i] != NULL; i++)
+		omap2_clkdm_allow_idle(clockdomains_omap[i]);
+
+	for (i = 0; clockdomains_omap[i] != NULL; i++)
+		omap2_clkdm_deny_idle(clockdomains_omap[i]);
+
+	for (i = 0; clockdomains_omap[i] != NULL; i++) {
+		val = omap2_clkdm_sleep(cd = clockdomains_omap[i]);
+		if (val == 0)
+			printk(KERN_INFO "Forced %s to sleep\n", cd->name);
+		else
+			printk(KERN_ERR "Failed with value %d for %s\n",
+								 val, cd->name);
+	}
+	for (i = 0; clockdomains_omap[i] != NULL; i++) {
+		val = omap2_clkdm_wakeup(cd = clockdomains_omap[i]);
+
+		if (val == 0)
+			printk(KERN_INFO "Forced %s to Wake up\n", cd->name);
+		else
+			printk(KERN_ERR "Failed with value %d for %s\n",
+								 val, cd->name);
+	}
+}
+EXPORT_SYMBOL(clock_domain_test);
Index: kernel-omap3/arch/arm/mach-omap2/clockdomains.h
===================================================================
--- kernel-omap3.orig/arch/arm/mach-omap2/clockdomains.h	2009-12-14 18:36:46.000000000 +0530
+++ kernel-omap3/arch/arm/mach-omap2/clockdomains.h	2009-12-17 13:05:53.000000000 +0530
@@ -25,18 +25,21 @@
 static struct clockdomain wkup_clkdm = {
 	.name		= "wkup_clkdm",
 	.pwrdm		= { .name = "wkup_pwrdm" },
+	.flags		= CLKDM_CAN_HWSUP_SWSUP,
 	.omap_chip	= OMAP_CHIP_INIT(CHIP_IS_OMAP24XX | CHIP_IS_OMAP3430),
 };

 static struct clockdomain prm_clkdm = {
 	.name		= "prm_clkdm",
 	.pwrdm		= { .name = "wkup_pwrdm" },
+	.flags		= CLKDM_CAN_HWSUP_SWSUP,
 	.omap_chip	= OMAP_CHIP_INIT(CHIP_IS_OMAP24XX | CHIP_IS_OMAP3430),
 };

 static struct clockdomain cm_clkdm = {
 	.name		= "cm_clkdm",
 	.pwrdm		= { .name = "core_pwrdm" },
+	.flags		= CLKDM_CAN_HWSUP_SWSUP,
 	.omap_chip	= OMAP_CHIP_INIT(CHIP_IS_OMAP24XX | CHIP_IS_OMAP3430),
 };

@@ -148,7 +151,7 @@
 static struct clockdomain mpu_34xx_clkdm = {
 	.name		= "mpu_clkdm",
 	.pwrdm		= { .name = "mpu_pwrdm" },
-	.flags		= CLKDM_CAN_HWSUP | CLKDM_CAN_FORCE_WAKEUP,
+	.flags		= CLKDM_CAN_HWSUP_SWSUP,
 	.clktrctrl_mask = OMAP3430_CLKTRCTRL_MPU_MASK,
 	.omap_chip	= OMAP_CHIP_INIT(CHIP_IS_OMAP3430),
 };
@@ -203,7 +206,7 @@
 static struct clockdomain core_l3_34xx_clkdm = {
 	.name		= "core_l3_clkdm",
 	.pwrdm		= { .name = "core_pwrdm" },
-	.flags		= CLKDM_CAN_HWSUP,
+	.flags		= CLKDM_CAN_HWSUP_SWSUP,
 	.clktrctrl_mask = OMAP3430_CLKTRCTRL_L3_MASK,
 	.omap_chip	= OMAP_CHIP_INIT(CHIP_IS_OMAP3430),
 };
@@ -211,7 +214,7 @@
 static struct clockdomain core_l4_34xx_clkdm = {
 	.name		= "core_l4_clkdm",
 	.pwrdm		= { .name = "core_pwrdm" },
-	.flags		= CLKDM_CAN_HWSUP,
+	.flags		= CLKDM_CAN_HWSUP_SWSUP,
 	.clktrctrl_mask = OMAP3430_CLKTRCTRL_L4_MASK,
 	.omap_chip	= OMAP_CHIP_INIT(CHIP_IS_OMAP3430),
 };
@@ -263,30 +266,35 @@
 static struct clockdomain dpll1_clkdm = {
 	.name		= "dpll1_clkdm",
 	.pwrdm		= { .name = "dpll1_pwrdm" },
+	.flags		= CLKDM_CAN_HWSUP_SWSUP,
 	.omap_chip	= OMAP_CHIP_INIT(CHIP_IS_OMAP3430),
 };

 static struct clockdomain dpll2_clkdm = {
 	.name		= "dpll2_clkdm",
 	.pwrdm		= { .name = "dpll2_pwrdm" },
+	.flags		= CLKDM_CAN_HWSUP_SWSUP,
 	.omap_chip	= OMAP_CHIP_INIT(CHIP_IS_OMAP3430),
 };

 static struct clockdomain dpll3_clkdm = {
 	.name		= "dpll3_clkdm",
 	.pwrdm		= { .name = "dpll3_pwrdm" },
+	.flags		= CLKDM_CAN_HWSUP_SWSUP,
 	.omap_chip	= OMAP_CHIP_INIT(CHIP_IS_OMAP3430),
 };

 static struct clockdomain dpll4_clkdm = {
 	.name		= "dpll4_clkdm",
 	.pwrdm		= { .name = "dpll4_pwrdm" },
+	.flags		= CLKDM_CAN_HWSUP_SWSUP,
 	.omap_chip	= OMAP_CHIP_INIT(CHIP_IS_OMAP3430),
 };

 static struct clockdomain dpll5_clkdm = {
 	.name		= "dpll5_clkdm",
 	.pwrdm		= { .name = "dpll5_pwrdm" },
+	.flags		= CLKDM_CAN_HWSUP_SWSUP,
 	.omap_chip	= OMAP_CHIP_INIT(CHIP_GE_OMAP3430ES2),
 };

@@ -314,7 +322,7 @@
  *
  */

-static struct clockdomain *clockdomains_omap[] = {
+struct clockdomain *clockdomains_omap[] = {

 	&wkup_clkdm,
 	&cm_clkdm,
@@ -342,7 +350,9 @@
 	&mpu_34xx_clkdm,
 	&neon_clkdm,
 	&iva2_clkdm,
+#ifndef CONFIG_ARCH_OMAP34XX
 	&gfx_3430es1_clkdm,
+#endif
 	&sgx_clkdm,
 	&d2d_clkdm,
 	&core_l3_34xx_clkdm,
Index: kernel-omap3/arch/arm/mach-omap2/Makefile
===================================================================
--- kernel-omap3.orig/arch/arm/mach-omap2/Makefile	2009-12-17 12:39:44.000000000 +0530
+++ kernel-omap3/arch/arm/mach-omap2/Makefile	2009-12-17 12:56:55.000000000 +0530
@@ -12,6 +12,9 @@
 # Power Domain Testcase for OMAP3
 obj-$(CONFIG_ARCH_OMAP3)                += pwr_dm_test.o

+# Clock Domain Testcase for OMAP3
+obj-$(CONFIG_ARCH_OMAP3)                += clk_dm_test.o
+
 obj-$(CONFIG_ARCH_OMAP2) += $(omap-2-3-common) $(prcm-common) $(clock-common)
 obj-$(CONFIG_ARCH_OMAP3) += $(omap-2-3-common) $(prcm-common) $(clock-common)

